unidler.py

```diff
@@ -6,36 +7,74 @@
 import ssl
 import sys
 
-# import kubernetes
-# import requests
+import kubernetes
+from kubernetes import client, config
+from kubernetes.client.models import (
+    V1beta1HTTPIngressPath,
+    V1beta1HTTPIngressRuleValue,
+    V1beta1IngressBackend,
+    V1beta1IngressRule,
+)
+
+
+IDLED = 'mojanalytics.xyz/idled'
+IDLED_AT = 'mojanalytics.xyz/idled-at'
+INGRESS_CLASS = 'kubernetes.io/ingress.class'
+UNIDLER = 'unidler'
 
 
 logging.basicConfig(level=os.environ.get('LOG_LEVEL', 'DEBUG'))
 log = logging.getLogger('unidler')
 
 
-class RequestHandler(BaseHTTPRequestHandler):
+def run(host='0.0.0.0', port=8080):
+    try:
+        config.load_incluster_config()
+    except:
+        config.load_kube_config()
 
-    def do_HEAD(self):
-        self.respond(200)
+    unidler = UnidlerServer((host, port), RequestHandler)
+    log.info(f'Unidler listening on {host}:{port}')
+    unidler.serve_forever()
+
+
+class UnidlerServer(ThreadingMixIn, HTTPServer):
+    pass
+
+
+class RequestHandler(BaseHTTPRequestHandler):
 
     def do_GET(self):
-        log.debug(self.requestline)
-        log.debug(self.headers)
+        hostname = self.headers.get('X-Forwarded-Host')
+
+        if not hostname:
+            return self.respond(200, 'OK')
 
-        self.cache_request()
-        self.mark_unidling()
-        # self.restore_replicas()
-        # self.restore_ingress()
-        # self.unmark_idled()
+        log.info(f'Received {self.requestline} for host {hostname}')
 
-        self.respond(200, 'OK')
+        try:
+            unidle_deployment(hostname)
+            log.info('Unidled {hostname}')
+            url = f'//{hostname}{self.path}'
+            body = f'Unidling, please <a href="{url}">try again</a>'
+            self.respond(503, body, {
+                'Content-Type': 'text/html',
+                'Refresh': f'5; {url}',
+                'Retry-After': 5})
 
-    def cache_request(self):
-        pass
+        except (DeploymentNotFound, IngressNotFound) as not_found:
+            log.error(not_found)
+            self.respond(404, not_found)
 
-    def mark_unidling(self):
-        ingress_host = self.headers['Host']
+        except kubernetes.client.rest.ApiException as error:
+            log.error(error)
+            self.respond(500, error)
+
+    def do_HEAD(self):
+        self.respond(200)
+
+    def do_POST(self):
+        self.do_GET()
```

Comment:
I'm not familiar with `BaseHTTPRequestHandler` but as far as I can tell from quickly reading the documentation when a `XXX` request arrives it calls `do_XXX()`.

So here we're only handling `GET`, `HEAD` and `POST` requests using the same handler, which unidle the deployment with host equal to whatever the value of `X-Forwarded-Host` is and would send a simple "HTML" document with a link and a refresh header in the response.

1) Do we want/makes sense to handle requests with potentially side effects (like `POST`)?
2) Wouldn't the refreshed request always be a `GET`? 
2a) and wouldn't click on the link in the document perform a `GET` request regardless?
3) If we decide to unidle on any possible request, don't know if redirect for non-`GET` requests would make sense
4) User may get confused or get errors if the `GET` redirect would cause errors, e.g. if `POST /foo/1/delete` gets redirected to `GET /foo/1/delete` which then responds `400`/`404`(a bit of a stretched example, but you got the idea).

Maybe not an actual problem and maybe you already thought about it but I thought it was worth comment about it (or I may miss something)

Reply:
Actually, to have transparent unidling, we need to cache the requests and replay them when the deployment is ready to receive them. That's coming later. To address each of your points:
1. We want to handle all requests
2. The refresh is probably not a good idea, I will remove it
3. ditto
4. The redirect happens after a few seconds, during which time the "Unidling .." message is shown. This is no good for API calls, of course.

Reply:
Regarding point 1 (handle all request methods), does it mean we have to define all the `do_*()` methods also for `DELETE`, `PATCH`, `PUT`, etc, etc?

Reply:
We will do, yes